### PR TITLE
 #726 adding "node_info" rpc endpoint

### PR DESCRIPTION
**Pull request by wolfposd**
https://github.com/tendermint/tendermint/issues/726

gives info about own node: moniker/pubkey/height
other nodes moniker/pubkey/height
own node: fast_sync state and blockstoreheight

```
{
  "jsonrpc": "2.0",
  "id": "",
  "result": {
    "self": {
      "moniker": "gogland1",
      "pub_key": {
        "type": "ed25519",
        "data": "76074E7B7C004816FBD7B0201FBA788425416DBA20E19332EC8D4D7DBE6E6290"
      },
      "height": 3975
    },
    "fast_sync": false,
    "block_store_height": 3974,
    "peers": [
      {
        "moniker": "mainnode",
        "pub_key": {
          "type": "ed25519",
          "data": "7B694F326EEFB363B91D57B33714B5281A75150FF53E73F0AB4A644A626EDCF4"
        },
        "height": 3975
      }
    ]
  }
}
```

### CODE REVIEW

File: rpc/core/types/responses.go [hunk 1]
```diff
@@ -60,6 +60,19 @@ type ResultNetInfo struct {
...
+type ResultNodeInfo struct {
+	Self NodeInfo `json:"self"`
+	NodeInfos []NodeInfo `json:"node_infos"`
+	FastSync bool `json:"fast_sync"`
+	BlockStoreHeight int `json:"block_store_height"`
+}
+
+type NodeInfo struct {
+	Moniker string `json:"moniker"`
+	Address crypto.PubKeyEd25519 `json:"pubkey"`
```
Comment (melekes) [inline rpc/core/types/responses.go]: `PubKey crypto.PubKey`

File: rpc/core/types/responses.go [hunk 2]
```diff
@@ -60,6 +60,19 @@ type ResultNetInfo struct {
 	Peers     []Peer   `json:"peers"`
 }
 
+type ResultNodeInfo struct {
+	Self NodeInfo `json:"self"`
+	NodeInfos []NodeInfo `json:"node_infos"`
```
Comment (melekes) [inline rpc/core/types/responses.go]: `Peers`

**Comment by melekes** [comment]
Nice job!

We need to think about our API though. Maybe this should be a part of the `net_info`?

**Comment by ebuchman** [comment]
Isn't this redundant with what's already provided by `/status` and `/net_info`?

eg. `/status`:

```
{
  "jsonrpc": "2.0",
  "id": "",
  "result": {
    "node_info": {
      "pub_key": "3AC44C66A4D2C9226D664D0825BE8DE9B52920883C88341CA14374621A0D560A",
      "moniker": "207.154.203.177",
      "network": "atlas",
      "remote_addr": "",
      "listen_addr": "192.168.1.100:46656",
      "version": "0.11.1",
      "other": [
        "wire_version=0.6.2",
        "p2p_version=0.5.0",
        "consensus_version=v1/0.2.2",
        "rpc_version=0.7.0/3",
        "tx_index=on",
        "rpc_addr=tcp://0.0.0.0:46657"
      ]
    },
    "pub_key": {
      "type": "ed25519",
      "data": "F5D0D60647A48580EBEA3DBBA93A48726FAA5C2F87E3447DE179DC89D40A7614"
    },
    "latest_block_hash": "78AA7E9AA7BBF943B60F21BBCCE5534F4859953C",
    "latest_app_hash": "60636319BBFBD43120EF89458B17AE32D898A9AB",
    "latest_block_height": 3280,
    "latest_block_time": 1507843035111000000,
    "syncing": false
  }
}
```

and `/net_info`:

```
{
  "jsonrpc": "2.0",
  "id": "",
  "result": {
    "listening": true,
    "listeners": [
      "Listener(@192.168.1.100:46656)"
    ],
    "peers": [
      {
        "node_info": {
          "pub_key": "C795B8FB0E885BDF48B4F7EBF0A604A551D40A2379B36B48DEBF43B5D9F2F1C3",
          "moniker": "104.131.86.166",
          "network": "atlas",
          "remote_addr": "104.131.86.166:46656",
          "listen_addr": "104.131.86.166:46656",
          "version": "0.11.0-31a7e2b3",
          "other": [
            "wire_version=0.6.2",
            "p2p_version=0.5.0",
            "consensus_version=v1/0.2.2",
            "rpc_version=0.7.0/3",
            "tx_index=on",
            "rpc_addr=tcp://0.0.0.0:46657"
          ]
        },

...
```

**Comment by ebuchman** [comment]
Oh I see - it doesn't tell you about the peer's height. You'd have to hit them for their /status to find out. Hmmm ...

File: rpc/core/types/responses.go [hunk 3]
```diff
@@ -60,6 +60,19 @@ type ResultNetInfo struct {
...
+type ResultNodeInfo struct {
+	Self NodeInfo `json:"self"`
+	FastSync bool `json:"fast_sync"`
+	BlockStoreHeight int `json:"block_store_height"`
+	NodeInfos []NodeInfo `json:"peers"`
+}
+
+type NodeInfo struct {
+	Moniker string `json:"moniker"`
+	Address crypto.PubKey `json:"pub_key"`
```
Comment (ebuchman) [inline rpc/core/types/responses.go]: Address -> PubKey

**Comment by ebuchman** [comment]
Thanks @wolfposd . Though I also just realized that the pubkey we use for ourselves and the pubkeys we're using for the peers aren't compatible. Our pubkey refers to our validator, while the peers pubkeys refer to the p2p layer. Recall that every node has two pubkeys - one for its validator, one for its p2p node. 

I would propose we drop the pubkey from this entirely and just use the Height and Moniker. Is that still useful?

**Comment by zramsay** [comment]
@wolfposd do you still want to pick this up?

**Comment by ebuchman** [comment]
We can't merge this as is because the pubkeys are inconsistent. 

Also it would probably make sense now to use the peer ID instead of the pubkey anyways (now that we have peer IDs on develop).  That would help avoid the pubkey confusion too. 

There's also a larger discussion here on whether this endpoint belongs in Tendermint. Can you clarify what you need this for ? Perhaps it's something better suited for a monitoring tool, eg. https://github.com/tendermint/tools/tree/master/tm-monitor (NOTE it may not be up to date, thats a separate issue) ? 

I'll close this in the meantime. Feel free to re open it rebased on develop and using IDs and a brief explanation of the use case.

**Pull request closed without merging**